pkg/codegen/jenny_crd_yaml.go

```diff
@@ -0,0 +1,213 @@
+package codegen
+
+import (
+	"bytes"
+	"fmt"
+	"path/filepath"
+
+	"cuelang.org/go/cue"
+	"cuelang.org/go/cue/ast"
+	"cuelang.org/go/encoding/openapi"
+	cueyaml "cuelang.org/go/pkg/encoding/yaml"
+	"github.com/grafana/codejen"
+	"github.com/grafana/grafana/pkg/kindsys"
+	"github.com/grafana/grafana/pkg/kindsys/k8ssys"
+	"github.com/grafana/thema"
+	goyaml "gopkg.in/yaml.v3"
+)
+
+// TODO this jenny is quite sloppy, having been quickly adapted from oasdk. It needs love
+
+// YamlCRDJenny generates a representation of a core structured kind in YAML CRD form.
+func YamlCRDJenny(path string) OneToOne {
+	return yamlCRDJenny{
+		parentpath: path,
+	}
+}
+
+type yamlCRDJenny struct {
+	parentpath string
+}
+
+func (yamlCRDJenny) JennyName() string {
+	return "YamlCRDJenny"
+}
+
+func (j yamlCRDJenny) Generate(decl *DefForGen) (*codejen.File, error) {
+	if !decl.IsCore() {
+		return nil, nil
+	}
+
+	props := decl.SomeDef.Properties.(kindsys.CoreProperties)
+	lin := decl.Lineage()
+
+	// We need to go through every schema, as they all have to be defined in the CRD
+	sch, err := lin.Schema(thema.SV(0, 0))
+	if err != nil {
+		return nil, err
+	}
+
+	resource := customResourceDefinition{
+		APIVersion: "apiextensions.k8s.io/v1",
+		Kind:       "CustomResourceDefinition",
+		Metadata: customResourceDefinitionMetadata{
+			Name: fmt.Sprintf("%s.%s", props.PluralMachineName, props.CRD.Group),
+		},
+		Spec: k8ssys.CustomResourceDefinitionSpec{
+			Group: props.CRD.Group,
+			Scope: props.CRD.Scope,
+			Names: k8ssys.CustomResourceDefinitionSpecNames{
+				Kind:   props.Name,
+				Plural: props.PluralMachineName,
+			},
+			Versions: make([]k8ssys.CustomResourceDefinitionSpecVersion, 0),
+		},
+	}
+	latest := lin.Latest().Version()
+
+	for sch != nil {
+		oapi, err := generateOpenAPI(sch, props)
+		if err != nil {
+			return nil, err
+		}
+
+		vstr := versionString(sch.Version())
+		if props.Maturity.Less(kindsys.MaturityStable) {
+			vstr = "v0-0alpha1"
+		}
+
+		ver, err := valueToCRDSpecVersion(oapi, vstr, sch.Version() == latest)
+		if err != nil {
+			return nil, err
+		}
+		if props.CRD.DummySchema {
+			ver.Schema = map[string]any{
+				"openAPIV3Schema": map[string]any{
+					"type": "object",
+					"properties": map[string]any{
+						"spec": map[string]any{
+							"type":                                 "object",
+							"x-kubernetes-preserve-unknown-fields": true,
```

Comment:
@toddtreece -- can you confirm this is the correct location?

Reply:
yep